[SPARK-16256][SQL][STREAMING] Added Structured Streaming Programming Guide

tdas opened the pull request:
Title defines all.

tdas committed:
Added Structured Streaming Programming Guide

SparkQA commented:
**[Test build #61380 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61380/consoleFull)** for PR 13945 at commit [`74108e5`](https://github.com/apache/spark/commit/74108e558b6d497546f7657971648d06881c7f58).

SparkQA commented:
**[Test build #61380 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61380/consoleFull)** for PR 13945 at commit [`74108e5`](https://github.com/apache/spark/commit/74108e558b6d497546f7657971648d06881c7f58).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class DeviceData(device: String, type: String, signal: Double, time: DateTime)`
  * `- The writer must do all the initialization (e.g. opening connections, starting a transaction, etc.) only when the `open` method is called. Be aware that, if there is any initialization in the class as soon as the object is created, then that initialization will happen in the driver (because that is where the instance is being created), which may not be what you intend.`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61380/
Test PASSed.

anabranch commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+# Overview
+Structured Streaming is a scalable and fault-tolerant stream processing engine 
+built on the Spark SQL engine. You can express your streaming computation by 
+thinking you are running a batch computation on a static dataset, and the 
```
I think we can work on the wording here a bit.

```
Structured Streaming is a scalable and fault-tolerant stream processing engine built on the Spark SQL engine. You can express your streaming computation the same way you would express a batch computation on a static dataset.The SparkSQL engine will take care of running it incrementally and continuously and updating the final result as streaming data continues to arrive.
```

anabranch commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+</div>
+</div>
+
+This `lines` DataFrame is like an unbounded table containing the streaming 
```
Change to:

> The `lines` DataFrame as defined above is represented as an unbounded table.

anabranch commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+
+This `lines` DataFrame is like an unbounded table containing the streaming 
+text data. This table contains one column of string named “value”, and each 
+line in the streaming text data is like a row in this table. Note, that this 
```
is like a > is represented as

anabranch commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+
+We have now set up the query on the streaming data. All that is left is to 
+actually start receiving data and computing the counts. To do this, we set it 
+up to output the counts to the console every time they are updated. In 
```
Think we can clean up these last two sentences.

> There are other details that will be mentioned on in the guide that are relevant for this problem like the specific checkpoint location where checkpoint data will be stored.

anabranch commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+</div>
+</div>
+
+Then, any lines typed in the terminal running the netcat server will be counted and printed on screen every second. It will look something like the following.
```
I don't see an example here... you might just be able to remove that last sentence?

koeninger replied:
is the intention here to follow with a cut & paste of terminal output?

tdas replied:
yep. added.

anabranch commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+
+  - *Complete Mode* - The entire updated Result Table will be written to the external storage.
+
+  - *Append Mode* - Only the new rows appended in the Result Table since the last trigger will be written to the external storage. This is applicable only on queries where existing rows in the Result Table is not expected to change.
```
> This is applicable on to queries where existing rows in the Result Table are not expected to change.

koeninger replied:
is not expected => are not expected

tdas replied:
fixed.

anabranch commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+see how this model handles event-time based processing and late arriving data.
+
+## Handling Event-time and Late Data
+Event-time is the time embedded in the data itself. For many applications, you may want to do operate using this event-time. For example, if you want to get the number of events generated by IoT devices every minute, then you probably want to use the time when the data was generated (that is, event-time in the data), rather than the time Spark receives them. This event-time is very naturally expr ...
```
replace
> you may want to do operate using this event-time.

with

> you may want to perform operations on event-time.

koeninger replied:
want to do operate => want to operate
on streaming dataset => on a streaming dataset

tdas replied:
fixed.

anabranch commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+Furthermore this model naturally handles data that has arrived later than expected based on its event-time. Since Spark is updating the Result Table, it has full control over updating/cleaning up the aggregates when there is late data. While not yet implemented in Spark 2.0 yet, event-time watermarking will be used to manage this data. These are explained in more details in the Window Operations  ...
+
+## Fault Tolerance Semantics
+Delivering end-to-end exactly-once semantics was one of key goals behind the design of Structured Streaming. To achieve that, we have designed the Structured Streaming sources, the sinks and the execution engine to reliably track the exact progress of the processing so that it can handle any kind of failure by restarting and/or reprocessing. Every streaming source is assumed to have offsets (simi ...
```
missing period at the end.

Could be good to define offsets just for people that might not be familiar.

koeninger replied:
If you aren't going to qualify exactly-once in the intro, need to do it here.  Producing into Kafka isn't exactly-once without additional work, for instance.

zsxwing commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+**Spark 2.0 is the ALPHA RELEASE of Structured Streaming** and the APIs are still experimental. In this guide, we are going to walk you through the programming model and the APIs. First, lets start with a simple example - a streaming word count. 
+
+# Quick Example
+Let’s say you want maintain a running word count of text data received from a data server listening on a TCP socket. Let’s see how you can express this using Structured Streaming. You can see the full code in Scala/Java/Python. And if you download Spark, you can directly run the example. In any case, let’s walk through the example step-by-step and understand how it is works. First, we have to imp ...
```
nit: want maintain -> want **to** maintain

zsxwing replied:
nit: understand how it ~~is~~ works. First

koeninger replied:
it is works => it works

tdas replied:
fixed.

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+[Dataset/DataFrame API](sql-programming-guide.html) in Scala, Java or Python to express streaming 
+aggregations, event-time windows, stream-to-batch joins, etc. The computation 
+is executed on the same optimized Spark SQL engine. Finally, the system 
+ensures end-to-end exactly-once fault-tolerance guarantees through 
```
End-to-end exactly-once sounds like over-promising.  Should probably define what the ends are, because destructive outputs can't be literally exactly-once in the face of network failures.

tdas replied:
ensures --> can ensure

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+processing model that is very similar to a batch processing model. You will 
+express your streaming computation as standard batch-like query as on a static 
+table, and Spark runs it as an *incremental* query on the *unbounded* input 
+table. Let’s understand this model in more details.
```
details => detail

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+table. Let’s understand this model in more details.
+
+## Basic Concepts
+Consider the input data stream as the “Input Table”. Every data items that is 
```
items => item

tdas replied:
fixed.

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+
+![Stream as a Table](img/structured-streaming-stream-as-a-table.png "Stream as a Table")
+
+A query on the input will generate the “Result Table”. Every trigger interval (say, every 1 second), new rows gets appended to the Input Table, which eventually updates the Result Table. Whenever the result table gets updated, we would want write the changed result rows to an external sink. 
```
new rows gets appended => new rows get appended

we would want write => we would want to write
(or just "we write")

tdas replied:
fixed.

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+
+The “Output” is defined as what gets written out to the external storage. The output can be defined in different modes 
+
+  - *Complete Mode* - The entire updated Result Table will be written to the external storage.
```
Need to clarify whether this duplicates entire result table to external storage repeatedly, i.e. at every trigger interval

tdas replied:
added something.

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+
+  - *Append Mode* - Only the new rows appended in the Result Table since the last trigger will be written to the external storage. This is applicable only on queries where existing rows in the Result Table is not expected to change.
+  
+  - *Update Mode* - Only the rows that were updated in the Result Table since the last trigger wil be written to the external storage (not available yet in Spark 2.0). Note that this is different from the Complete Mode in that this mode does not output the rows that are not changed.
```
wil be written => will be written

tdas replied:
fixed.

holdenk commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+
+- **Foreach sink** - Runs arbitrary computation on the records in the output. See later in the section for more details.
+
+- **Console sink (for debugging)** - Prints the output to the console/stdout every time there is a trigger. Both, Append and Complete output modes, are supported.
```
We should add a note (even though its marked as for debugging) that the entire slice is collected to the driver even if there is a limit (after SPARK-16020)

tdas replied:
fixed.

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+## Handling Event-time and Late Data
+Event-time is the time embedded in the data itself. For many applications, you may want to do operate using this event-time. For example, if you want to get the number of events generated by IoT devices every minute, then you probably want to use the time when the data was generated (that is, event-time in the data), rather than the time Spark receives them. This event-time is very naturally expr ...
+
+Furthermore this model naturally handles data that has arrived later than expected based on its event-time. Since Spark is updating the Result Table, it has full control over updating/cleaning up the aggregates when there is late data. While not yet implemented in Spark 2.0 yet, event-time watermarking will be used to manage this data. These are explained in more details in the Window Operations  ...
```
yet implemented in Spark 2.0 yet => yet implemented in Spark 2.0

tdas replied:
fixed.

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+
+## Fault Tolerance Semantics
+Delivering end-to-end exactly-once semantics was one of key goals behind the design of Structured Streaming. To achieve that, we have designed the Structured Streaming sources, the sinks and the execution engine to reliably track the exact progress of the processing so that it can handle any kind of failure by restarting and/or reprocessing. Every streaming source is assumed to have offsets (simi ...
+for track the read position in the stream. The engine uses checkpointing and write ahead logs to record the offset range of the data being processed in each trigger. The streaming sinks are designed to be idempotent for handling reprocessing. Together, Structured Streaming can ensure **end-to-end exactly-once semantics** under any failure.
```
for track => to track

Again, "under any failure" is over-promising.

tdas replied:
I qualified the statement as "Together, using replayable sources and idempotant sinks, Structured Streaming can ensure..."

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+for track the read position in the stream. The engine uses checkpointing and write ahead logs to record the offset range of the data being processed in each trigger. The streaming sinks are designed to be idempotent for handling reprocessing. Together, Structured Streaming can ensure **end-to-end exactly-once semantics** under any failure.
+
+# API using Datasets and DataFrames
+Since Spark 2.0, DataFrames and can represent of static, bounded data, as well as streaming, unbounded data. Similar to static Datasets/DataFrames, you can use the common entry point  SparkSession to create straming DataFrames/Datasets from streaming sources, and apply the same operations on them as static DataFrames/Datasets. If you are not familiar with Datasets/DataFrames, you are strongly adv ...
```
DataFrames and can represent of static => Datasets and DataFrames can represent static
(I'm assuming that was the intention)

create straming DataFrames => create streaming DataFrames

tdas replied:
fixed.

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+Since Spark 2.0, DataFrames and can represent of static, bounded data, as well as streaming, unbounded data. Similar to static Datasets/DataFrames, you can use the common entry point  SparkSession to create straming DataFrames/Datasets from streaming sources, and apply the same operations on them as static DataFrames/Datasets. If you are not familiar with Datasets/DataFrames, you are strongly adv ...
+
+## Creating streaming DataFrames and streaming Datasets
+Streaming DataFrames can created through the DataStreamReader interface `SparkSession.readStream()`. Similar to the read interface for creating static DataFrame, you can specify the details of the source - data format, schema, options, etc. In Spark 2.0, there are a few built-in sources.
```
DataFrames can created through => DataFrames can be created through

tdas replied:
fixed.

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+## Creating streaming DataFrames and streaming Datasets
+Streaming DataFrames can created through the DataStreamReader interface `SparkSession.readStream()`. Similar to the read interface for creating static DataFrame, you can specify the details of the source - data format, schema, options, etc. In Spark 2.0, there are a few built-in sources.
+
+  - **File sources** - Reads files written in a directory as a stream of data. Supported file formats are text, csv, json, parquet. See the docs of the DataStreamReader interface for more up-to-date list, and supported options for each file format. Note that the files needs to be atomically put in the directory. In most file systems, this means that the files should be moved atomically into the d ...
```
for more up-to-date list, => for a more up-to-date list

files needs to be atomically => files need to be atomically

tdas replied:
fixed.

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+
+  - **File sources** - Reads files written in a directory as a stream of data. Supported file formats are text, csv, json, parquet. See the docs of the DataStreamReader interface for more up-to-date list, and supported options for each file format. Note that the files needs to be atomically put in the directory. In most file systems, this means that the files should be moved atomically into the d ...
+
+  - **Socket source (for testing)** - Reads UTF8 text data from a socket connection. The connection is only on the server. Note that this should be used only for testing as this does not provide end-to-end fault-tolerance guarantees. 
```
I'm not sure what "The connection is only on the server" means in this case.

tdas replied:
Fixed.

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+    .option("port", 9999)
+    .load()
+
+socketDF.isStreaming    // Returns true for DataFrames that has a streaming source
```
DataFrames that has => DataFrames that have

sethah commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+</div>
+</div>
+
+Using Foreach
```
Is this supposed to be a header? It doesn't display correctly the way it is now.

tdas replied:
Fixed.

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+    .option("port", 9999)
+    .load();
+
+socketDF.isStreaming();    // Returns True for DataFrames that has a streaming source
```
DataFrames that has => DataFrames that have

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+    .option("port", 9999) \
+    .load()
+
+socketDF.isStreaming()    # Returns True for DataFrames that has a streaming source
```
DataFrames that has => DataFrames that have

sethah commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+</div>
+
+Using Foreach
+The `foreach` operation allows arbitrary operations to be computer on the output data. As of Spark 2.0, this is available only for Scala and Java, not Python. To use this, you will have to implement the interface `ForeachWriter`, which has methods that gets called whenever there is a sequence of rows generated as output after a trigger. It is important note the following
```
"to be computer" -> "to be computed"

koeninger replied:
operations to be computer => operations to be computed
 methods that gets called =>  methods that get called 
is important note the => is important to note the

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+</div>
+</div>
+
+These examples generate streaming DataFrames that are untyped, that is, the schema of the DataFrame is not checked at compile time, only checked in runtime when the query is submitted. Some operations like `map`, `flatMap`, etc. need the type to be known at compile time. To do those, you can convert these untyped streaming DataFrames to typed streaming Datasets using the same methods as static Da ...
```
only checked in runtime => only checked at runtime

I think at least one example of compile time checking on a streaming Dataset would be good.

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+These examples generate streaming DataFrames that are untyped, that is, the schema of the DataFrame is not checked at compile time, only checked in runtime when the query is submitted. Some operations like `map`, `flatMap`, etc. need the type to be known at compile time. To do those, you can convert these untyped streaming DataFrames to typed streaming Datasets using the same methods as static Da ...
+
+## Operations on streaming DataFrames/Datasets
+You can apply all kinds of streaming DataFrames/Datasets - ranging for untyped, SQL-like operations (e.g. select, where, groupBy), to typed RDD-like operations (e.g. map, filter, flatMap). See the SQL programming guide for more details. Let’s take a look at a few example operations that you can use.
```
all kinds of streaming DataFrames/Datasets - ranging for =>
all kinds of operations to streaming DataFrames/Datasets - ranging from

tdas replied:
fixed

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+You can apply all kinds of streaming DataFrames/Datasets - ranging for untyped, SQL-like operations (e.g. select, where, groupBy), to typed RDD-like operations (e.g. map, filter, flatMap). See the SQL programming guide for more details. Let’s take a look at a few example operations that you can use.
+
+### Basic Operations - Selection, Projection, Aggregation
+Most of the common operations on DataFrame/Dataset operations are supported for streaming. The few operations that are not supported are discussed later in this section.
```
operations on DataFrame/Dataset operations => DataFrame/Dataset operations

tdas replied:
fixed.

sethah commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+
+- The writer must do all the initialization (e.g. opening connections, starting a transaction, etc.) only when the `open` method is called. Be aware that, if there is any initialization in the class as soon as the object is created, then that initialization will happen in the driver (because that is where the instance is being created), which may not be what you intend.
+
+- `version` and `partition` are two parameter in the `open` that uniquely represents a set of rows that needs to be pushed out. `version` is monotonically increasing id that increases with every trigger. `partition` is an id that represents a partition of the output, since the output is distributed and will be processed on multiple executors.
```
"are two parameter ... that uniquely represents" -> "are two parameters ... that uniquely represent"
"`version` is monotonically" -> "`version` is a monotonically"

koeninger replied:
two parameter in the `open` that uniquely represents => two parameters in the `open` that uniquely represent

`version` is monotonically increasing id => `version` is a monotonically increasing id

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+{% highlight scala %}
+case class DeviceData(device: String, type: String, signal: Double, time: DateTime)
+
+val df: DataFrame = … // streaming DataFrame with IOT device data with schema { device: string, type: string, signal: double, time: string }
```
Is there a way to make this at least potentially runnable, rather than an ellipsis placeholder?

tdas replied:
kind of hard to keep the code snippet concise AND runnable. creating the socket stream everytime is ugly and verbose. I am thinking that we should have something simple like `spark.range()` but for streaming DFs.... something that generates random data

sethah commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+
+- `version` and `partition` are two parameter in the `open` that uniquely represents a set of rows that needs to be pushed out. `version` is monotonically increasing id that increases with every trigger. `partition` is an id that represents a partition of the output, since the output is distributed and will be processed on multiple executors.
+
+- `open` can use the `version` and `partition` to choose whether it need to write the sequence of rows. Accordingly, it can return `true` (proceed with writing), or `false` (no need to write). If the `false` is returned, then `write` will not be called on any row. For example, after a partial failure, so partitions of the failed trigger may have already been committed to a database. Based on meta ...
```
"whether it need" -> "whether it needs"
"If the `false`" -> "If `false`"
"so partitions" -> "some partitions"

"been committed and" ...? the end of this bullet seems to be missing

koeninger replied:
whether it need to write => whether it needs to write

If the `false` is returned => If `false` is returned

partitions that have already been committed and  =>  incomplete sentence?

tdas replied:
fixed

sethah commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+
+- `open` can use the `version` and `partition` to choose whether it need to write the sequence of rows. Accordingly, it can return `true` (proceed with writing), or `false` (no need to write). If the `false` is returned, then `write` will not be called on any row. For example, after a partial failure, so partitions of the failed trigger may have already been committed to a database. Based on meta ...
+
+- Whenever `open` is called, `close` will also be called (unless the JVM exits due to some error). This is true even if `open` returns false. If there is any error in processing and writing the data, `close` will be called with the error. It is the your responsibilty to clean up state (e.g. connections, transactions, etc.) that have been created in `open` such that there are no resource leaks.
```
"It is the your" -> "It is your"

koeninger replied:
It is the your responsibilty => It is your responsibility

tdas replied:
fixed.

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+</div>
+
+### Window Operations on Event Time
+Aggregations over a sliding window is very easy. Applying a sliding window is simply a special type of grouping - each time window is a group and each row can belong to multiple windows/groups. Hence aggregating over sliding windows is same as aggregating over this special type of grouping. This is illustrated below. Imagine the quick example is modified and the stream contains lines along with t ...
```
Aggregations over a sliding window is very easy => Aggregations over a sliding window are very easy

Hence aggregating over sliding windows is same as => Hence aggregating over sliding windows is the same as

we want to count windows within 10 minute windows => we want to count words within 10 minute windows

"combined group (window, cat)" is a little confusing because it's one abstract and one concrete. I'd use both abstract (window, word) or both concrete (12:00-12:10, cat)

tdas replied:
reworded

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+
+![Handling Late Data](img/structured-streaming-late-data.png)
+
+### Joins Operations
```
Join Operations

tdas replied:
fixed

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+
+{% highlight scala %}
+// Number of events in every 1 minute time windows
+df.groupBy(window(time, "1 minute"))
```
I know dean and others commented on it, but seeing stringly-typed time definitions is unfortunate IMHO.

tdas replied:
keeps API and code snippets consistent across Java/scala/python. i agree its not the most optimized for scala users, but we are targeting getting convenience and consistency across languages over optimizing for one language.

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+If you try any of these operations, you will see an AnalysisException like "operation XYZ is not supported with streaming DataFrames/Datasets".
+
+## Starting Streaming Queries
+Once you have defined the final result DataFrame/Dataset, all that is left is for you start the StreamingQuery. To do that, you have to use the `DataStreamWriter`, through you have to specify the following. 
```
through you have to specify => through which you have to specify
(I'm assuming)

tdas replied:
fixed.

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+
+- *Query name:* Optionally, specify a unique name of the query for identification.
+
+- *Trigger interval:* Optionally, specify the trigger interval. If its not specified, the system will check for availability of new day as soon as the previous processing has completed. If a trigger time is missed because the previous processing has not completed, then the system will attempt to trigger at the next trigger point, not immediately after the processing has completed.
```
availability of new day => availability of new data

tdas replied:
fixed.

sethah commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+
+- The writer must be serializable, as it will be serialized and sent to the executors for execution.
+
+- All the three methods, `open`, `write` and `close` will be called on the executors.
```
I think `write` should actually be `process`, looking at the definition of `ForeachWriter`.

tdas replied:
my bad! fixed.

koeninger commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+Finally, for asynchronous monitoring of streaming queries, you can create and attach a StreamingQueryListener, which will give you regular callback-based updates when queries are started and terminated.
+
+## Recovering from Failures with Checkpointing 
+In case of a failure or intentional shutdown, you can recover the previous progress and state of a previous query, and continue where it left off. This is done using checkpointing and write ahead logs. You can configure query with a checkpoint location, and the query will save all the progress information (i.e. range of offsets processed in each trigger), and the running aggregates (e.g. word cou ...
```
can configure query with => can configure a query with

Should probably clarify what the checkpoint limitations are on code change.

tdas replied:
This is still fuzzy actually. I have to update in the future once we have the behavior more well defined. :(

ahmed-mahran commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+text data. This table contains one column of string named “value”, and each 
+line in the streaming text data is like a row in this table. Note, that this 
+is not currently receiving any data as we are just setting up the 
+transformation, and have not yet started it. Next, we have used to built-in 
```
Next, we have used **to** built-in => Next, we have used **the** built-in

ahmed-mahran commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,888 @@
+<div class="codetabs">
+<div data-lang="scala"  markdown="1">
+
+{% highlight java %}
```
{% highlight scala %}

tdas replied:
fixed

tdas committed:
Added links and addressed comments

tdas committed:
Merge remote-tracking branch 'apache-github/master' into SPARK-16256

tdas committed:
Addressed more comments

tdas commented:
Thank you very much everyone for the detailed review! I am really thankful you caught so many issues that I missed in my first pass. I have addressed your comments as well as more comments I have received offline.  

In the interest of Spark 2.0 release, I am going to prioritize merging this PR. If there are outstanding issues, lets solve them in follow up PRs. I am sure that we can improve this draft by a lot with everyone's contributions.

SparkQA commented:
**[Test build #61455 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61455/consoleFull)** for PR 13945 at commit [`78223a4`](https://github.com/apache/spark/commit/78223a4cdb55e035536ca4bf810cd49a7fdd4f49).

SparkQA commented:
**[Test build #61455 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61455/consoleFull)** for PR 13945 at commit [`78223a4`](https://github.com/apache/spark/commit/78223a4cdb55e035536ca4bf810cd49a7fdd4f49).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61455/
Test PASSed.

ScrapCodes commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,1156 @@
+
+Now consider what happens if one of the events arrives late to the application.
+For example, a word that was generated at 12:04 but it was received at 12:11. 
+Since this windowing is based on the time in the data, the time 12:04 should considered for windowing. This occurs naturally in our window-based grouping --the late data is automatically placed in the proper windows and the correct aggregates updated as illustrated below.
```
Couple of minor corrections.
1) the time 12:04, should *be* considered for windowing. 
2) grouping the late data

ScrapCodes commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,1156 @@
+    .readStream() \
+    .option("sep", ";") \
+    .schema(userSchema) \
+    .csv("/path/to/directory")    # Equivalent to format("cv").load("/path/to/directory")
```
format("csv")

ScrapCodes commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,1156 @@
+   .format("memory")
+   .start();
+
+spark.sql("select * from aggregates).show();   // interactively query in-memory table
```
"select * from aggregates"

ScrapCodes commented on docs/structured-streaming-programming-guide.md:
```diff
@@ -0,0 +1,1156 @@
+   .format("memory")\
+   .start()
+
+spark.sql("select * from aggregates).show()   # interactively query in-memory table
```
"select * from aggregates" ending quote is missing.

tdas commented:
@ScrapCodes thanks for catching those. I will update them in a follow up PR. I am merging this as is to master and 2.0 in the interest of making it to Spark 2.0 RC2

tdas commented:
I have opened up another PR #13978 with left over fixes.
